### PR TITLE
hal: create list interface

### PR DESCRIPTION
**Pull request by badochov**
JIRA: RTOS-724

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [ ] Tested by hand on: (list targets here).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing linter checks and tests passed.
- [ ] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: hal/list.h [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+			(*list) = NULL; \
+		else { \
+			t->prev->next = t->next; \
+			t->next->prev = t->prev; \
+			if (t == (*list)) \
+				(*list) = t->next; \
+		} \
+		t->next = NULL; \
+		t->prev = NULL; \
+	} while (0)
```
Comment (agkaminski) [inline hal/list.h:51]: No new line at the end of the file, this time even gcc is complaining

File: hal/list.h [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+ * Author: Hubert Badocha
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+
+/* list has to be a struct with pointer fields prev and next to another list */
+
```
Comment (agkaminski) [inline hal/list.h:18]: Missing header guard

File: hal/list.h [hunk 3]
```diff
@@ -0,0 +1,51 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Operating system kernel
+ *
+ * List
+ *
+ * Copyright 2023 Phoenix Systems
+ * Author: Hubert Badocha
```
Comment (agkaminski) [inline hal/list.h:9]: This implementation was adapted from elsewhere, would be nice to mention the OG too

File: hal/list.h [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+		if ((t->next == t) && (t->prev == t)) \
+			(*list) = NULL; \
+		else { \
+			t->prev->next = t->next; \
+			t->next->prev = t->prev; \
+			if (t == (*list)) \
+				(*list) = t->next; \
+		} \
+		t->next = NULL; \
+		t->prev = NULL; \
```
Comment (agkaminski) [inline hal/list.h:50]:
```suggestion
		if ((t) == NULL) { \
			break; \
		} \
		if (((t)->next == t) && ((t)->prev == t)) { \
			(*(list)) = NULL; \
		}
		else { \
			(t)->prev->next = (t)->next; \
			(t)->next->prev = (t)->prev; \
			if ((t) == (*(list))) \
				(*(list)) = (t)->next; \
		} \
		(t)->next = NULL; \
		(t)->prev = NULL; \
```

File: hal/sparcv8leon3/gaisler/irqmp.c [hunk 5]
```diff
@@ -86,7 +53,7 @@ extern unsigned int _end;
 
 struct {
 	volatile u32 *int_ctrl;
-	spinlock_t spinlocks[SIZE_INTERRUPTS];
+	spinlock_t spinlock_common[SIZE_INTERRUPTS];
```
Comment (lukileczo) [inline hal/sparcv8leon3/gaisler/irqmp.c:56]: Renamed a little bit too much (same in `irqamp.c`)

File: hal/list.h [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+			break; \
+		} \
+		(t)->prev = (*(list))->prev; \
+		(*(list))->prev->next = (t); \
+		(t)->next = (*(list)); \
+		(*(list))->prev = (t); \
+	} while (0)
+
+
+#define hal_listRemove(list, t) \
```
Comment (lukileczo) [inline hal/list.h:41]: Remove macro not updated to upper case

**Review by lukileczo — APPROVED** [review]
LGTM

**Review by agkaminski — APPROVED** [review]

**Pull request merged**